### ISSUE TITLE
Warn about variables that have no actual effect

### ISSUE THREAD

**Issue by forthrin** [rubocop/rubocop#14666]
## Is your feature request related to a problem? Please describe.

```ruby
def foo
  prev = nil
  [1, 2, 3].each do
    puts it
    prev = it
  end
end

foo
```

Here, `prev` is of no consequence to the program, ie.

- Not used for decision-making
- Not copied to other variables
- Not sent to other functions
- Not printed to the terminal
- Not available outside its scope

## Describe the solution you'd like

```Variable `prev` has no actual effect and can be removed [[Correctable] Lint/InconsequentialVariable]```

## Describe alternatives you've considered

None come to mind.

## Additional context

Nothing comes to mind.

**Comment by koic** [comment]
It seems that the existing `Lint/UselessAssignment` cop should be extended rather than introducing a new cop like `Lint/InconsequentialVariable`.

**Comment by forthrin** [comment]
Sounds reasonable and analogous to how other cops have nuanced variations within their jurisdiction.

**Comment by forthrin** [comment]
@koic: Do you have a viable strategy for this one?

**Comment by forthrin** [comment]
@koic: Had AI do a cost/benefit-analysis. Could we start with a low hanging fruit implementation and see how that goes?

| # | Implementation                                                                  | Cost        | Benefit    |
|--:|---------------------------------------------------------------------------------|-------------|------------|
| 1 | Detect local variables assigned but never read within the same scope            | Low         | High       |
| 2 | Detect reads in expressions, conditions, interpolation, indexing, and arguments | Low         | High       |
| 3 | Add autocorrection, preserving right-hand sides that may have side effects      | Medium      | High       |
| 4 | Handle repeated assignments when no assigned value is subsequently read         | Medium      | High       |
| 5 | Track assignments and reads across `if`, `case`, loops, and early returns       | Medium–High | High       |
| 6 | Handle simple aliases, parallel assignments, and destructuring                  | Medium      | Medium     |
| 7 | Detect variables captured by blocks, lambdas, `Proc`s, and nested scopes        | High        | Medium     |
| 8 | Track values escaping through arguments, yields, storage, or returns            | High        | Low–Medium |
| 9 | Handle `binding`, `eval`, reflection, metaprogramming, and dynamic code         | Very high   | Low        |

Note: Steps 6–9 have diminishing returns and substantially increase false-positive risk.
